libraries/Microsoft.Bot.Builder.Dialogs.Adaptive.Testing/DialogInspector.cs

```diff
@@ -0,0 +1,219 @@
+        {
+            // This class just lets you load & save memory in parallel
+            var botStateSet = new BotStateSet();
+
+            // Some of the memory scopes expect to find things like storage in the turn state
+            foreach (var pair in InitialTurnState)
+            {
+                context.TurnState.Set(pair.Key, pair.Value);
+            }
+
+            // register DialogManager with TurnState.
+            context.TurnState.Set(this);
+
+            // At a minimum you need ConversationState. UserState is optional
+            if (ConversationState == null)
+            {
+                ConversationState = context.TurnState.Get<ConversationState>() ?? throw new NullReferenceException(nameof(ConversationState));
+            }
+            else
+            {
+                context.TurnState.Set(ConversationState);
+            }
+
+            // Add conversation state & user state to our parallel class
+            botStateSet.Add(ConversationState);
+            if (UserState == null)
+            {
+                UserState = context.TurnState.Get<UserState>();
+            }
+            else
+            {
+                context.TurnState.Set(UserState);
+            }
+
+            if (UserState != null)
+            {
+                botStateSet.Add(UserState);
+            }
+
+            // get dialog stack 
+            var dialogsProperty = ConversationState.CreateProperty<DialogState>(_dialogStateProperty);
+            var dialogState = await dialogsProperty.GetAsync(context, () => new DialogState(), cancellationToken).ConfigureAwait(false);
+
+            // Create DialogContext
+            var dc = new DialogContext(Dialogs, context, dialogState);
+
+            // promote initial TurnState into dc.services for contextual services
+            foreach (var service in dc.Services)
+            {
+                dc.Services[service.Key] = service.Value;
+            }
+
+            // map TurnState into root dialog context.services
+            foreach (var service in context.TurnState)
+            {
+                dc.Services[service.Key] = service.Value;
+            }
+
+            // get the DialogStateManager configuration
+            // - this configures all of the memory scopes and makes sure all of their memory has been loaded.
+            var dialogStateManager = new DialogStateManager(dc, StateConfiguration);
+            await dialogStateManager.LoadAllScopesAsync(cancellationToken).ConfigureAwait(false);
+            dc.Context.TurnState.Add(dialogStateManager);
+
+            // Find the DC for the active dialog
+            var activeDc = GetActiveDialogContext(dc);
+
+            // save all state scopes to their respective botState locations.
+            await dialogStateManager.SaveAllChangesAsync(cancellationToken).ConfigureAwait(false);
+
+            inspector(activeDc);
```

Comment:
>inspector(activeDc); [](start = 12, length = 20)

Why is this between these two save state calls? The call above is pointless as nothing would have changed as we didn't call any dialogs.

Reply:
Fixed per suggestion.